### ISSUE TITLE
Use of filter_input()

### ISSUE THREAD

**Issue by petenelson** [10up/Engineering-Best-Practices#132]
I've used Code Climate for several personal projects and something that is recommended frequently is using filter_input() rather than accessing $_POST and $_GET directly.  I didn't see anything in our best practices, so this might be something we could include.

**Comment by Jaace** [comment]
I like this idea -- WordPress-VIP PHPCS highlights the access of these superglobals directly - it would be nice to avoid that issue. https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/wiki/Flagged-superglobal-usage-in-WordPress-VIP

**Comment by hamptonpaulk** [comment]
This is also an interesting note on the docs page. Saves a good bit of checking.
http://php.net/manual/en/function.filter-input.php#115086

**Comment by morganestes** [comment]
I agree that it's a good idea to have something in our docs about not using the values from `$_GET`/`$_POST`/`$_REQUEST` without some sort of sanitization/validation, but we need to be cautious about the wording. Simply stating that it's not a good idea with some references (these links are great ones) should do the trick, but much more than that can turn into a lengthy discussion on security that will probably never be complete.

My concern is related to the PHP manual on `filter_input()`, and the comment linked to by @hamptonpaulk: it glosses over the fact that calling `filter_input(INPUT_GET, 'a')` does absolutely **no filtering** on `$_GET['a']`. The commenter states that `FILTER_REQUIRE_SCALAR` is the default, but the truth is that `FILTER_UNSAFE_RAW` is the default, and it provides no filtering, which leads to a false sense of security.

If we do state a preference for `filter_input()`, we should highlight that it needs a valid filter (and flags) to be useful, and also provide concrete examples of which flags to use for the most common use cases we experience in our work. We also need some good reasons why its use is prefered to core sanitization functions. 

A couple of things to note: 
- It's interesting that there are only a [handful of tickets](https://core.trac.wordpress.org/search?q=filter_input) in WP Core related to `filter_input()`, and most mentions are removing or recommending against its use.
- The [Plugins Handbook](https://developer.wordpress.org/plugins/security/securing-input/#sanitizing) has a list of sanitization functions that are useful, with practical examples of their use.

**Comment by allan23** [comment]
Also see: https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/issues/512

**Comment by petenelson** [comment]
Adding this here as some possible example code... implemented this in a project today to save HTML on a user profile, callbacks can be used to call WordPress sanitize functions.

```
$other_content = filter_input( INPUT_POST, 'user-profile-other', FILTER_CALLBACK, array( 'options' => 'filter_html_content' ) );

update_user_meta( $user_id, 'user-profile-other', $other_content );

/**
 * Callback for filter_input to filter content using wp_kses via the user_description allowed_html
 *
 * @param  string   $content   the unfiltered content
 * @return string              filtered content
 */
function filter_html_content( $content ) {
    return wp_kses( $content, wp_kses_allowed_html( 'user_description' ) );
}
```

**Comment by tlovett1** [comment]
I'm not completely sold on `filter_input`. What if I just want to check if a `$_GET` variable is set? Should I use `filter_input`?

**Comment by Camwyn** [comment]
When someone next creates a PR to change the wording, can someone please close the `absint()` in the very first example under "Input Validation and Sanitization"?

**Comment by eugene-manuilov** [comment]
@tlovett1 if you just want to check if a variable exists, then you just need to use `isset` or `empty` functions. But if you need to get a value of a variable, then it is better to use `filter_input` function, because it gives you better instruments to check, sanitize and receive values.

**Comment by tlovett1** [comment]
@petenelson that seems like a lot of code to sanitize a global variable. @eugene-manuilov is it really worth using this when it makes WP's sanitization/escaping/validation library difficult to use?

**Comment by kkmuffme** [comment]
@tlovett1 you need to use `filter_has_var` to check if it's set or not
